fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/predicates/OrPredicate.java

```diff
@@ -107,6 +113,140 @@ public OrPredicate withChildren(final Iterable<? extends QueryPredicate> newChil
+        final ImmutableSet.Builder<RangeConstraints> rangesSet = ImmutableSet.builder();
+
+        for (final var child : predicate.getChildren()) {
+            final var rangesBuilder = RangeConstraints.newBuilder();
+            if (child instanceof ValuePredicate) {
+                final var valuePredicate = (ValuePredicate)child;
+                if (!rangesBuilder.addComparisonMaybe(valuePredicate.getComparison())) {
+                    return Optional.empty();
+                }
+            } else if (child instanceof ValueWithRanges) {
+                rangesSet.addAll(((ValueWithRanges)child).getRanges());
+            } else {
+                // unknown child type.
+                return Optional.empty();
+            }
+
+            final var range = rangesBuilder.build();
+            if (range.isEmpty()) {
+                return Optional.empty();
+            }
+            rangesSet.add(range.get());
+        }
+
+        return Optional.of(ValueWithRanges.constraint(value, rangesSet.build()));
+    }
+
+    /**
+     * Checks for implication with a match candidate predicate by constructing a disjunction set of compile-time
+     * ranges of the {@code this} and the candidate predicate and, if the construction is possible, matches them.
+     * Matching the disjunction sets works as the following:
+     * <br>
+     * given an LHS that is: range(x1,x2) ∪ range(x3, x4) ∪ range(x5, x6) and RHS that is range(y1, y2) ∪ range (y3, y4):
+     * - each range in the LHS must find a companion range in RHS that implies it, if not, we reject the candidate predicate.
+     * <br>
+     * - if each companion range is _also_ implied by the LHS range, we have a match that does not require any compensation.
+     * <br>
+     * - otherwise, we match with a compensation that is effectively the reapplication of the entire LHS on top.
+     * <br>
+     * <b>example 1:</b>
+     * - LHS: range(3, 10) ∪ range(15, 20), range (50, 60)
+     * - RHS: range(0, 50) ∪ range (51,200)
+     * result:
+     * - match with {@code this} applied as a residual.
+     * <br>
+     * <b>example 2:</b>
+     * - LHS: range(3,10) ∪ range(15,20)
+     * - RHS: range(15,20) ∪ range(3,10)
+     * result:
+     * - exact match (no compensation required)
+     * <br>
+     * <b>example 3:</b>
+     * - LHS: range(3,10) ∪ range(15,20)
+     * - RHS: range(3,17)
+     * result:
+     * - no match.
+     *
+     * @param aliasMap the current alias map.
+     * @param candidatePredicate another predicate to match.
+     * @return optional match mapping.
+     */
+    @Nonnull
+    @Override
+    public Optional<PredicateMultiMap.PredicateMapping> impliesCandidatePredicate(@NonNull final AliasMap aliasMap, @Nonnull final QueryPredicate candidatePredicate) {
+        final var valueWithRangesMaybe = toValueWithRangesMaybe(this);
```

Comment:
Not good I think, this should at least be memoized. But this is also in contrast to how `SelectExpression` currently groups their predicates.

Reply:
Memoization done, will create a follow-up ticket to have a unified interface for `OrPredicate`/`AndPredicate` with `ValueWithRanges`.

Reply:
Here is the ticket: #2009

Reply:
Thank you!